### PR TITLE
[M] Migrate cloud_registration_spec and localization_spec to Java (ENT-5258)

### PR DESCRIPTION
**Pull request by bakajstep**

### CODE REVIEW

File: src/main/java/org/candlepin/resource/ConsumerResource.java
Comment (Ceiu) [inline src/main/java/org/candlepin/resource/ConsumerResource.java:657]: I *really* don't want to see all of the various headers and parameters start showing up in methods like this. Do we have a better way of doing the OpenAPI stuff to avoid this?

File: src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java
Comment (Ceiu) [inline src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java:297]: If a linter told you to remove that indentation, it was wrong. If anything, this should be re-written as a `doReturn(...).when(...)` type of mock, as it might help with readability in both testing and as far as the linter goes.

File: src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java
Comment (Ceiu) [inline src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java:536]: Carry-on line should be indented

**Review by Ceiu — CHANGES_REQUESTED** [review]
I'm hoping we have something better we can do with the headers and parameters so we don't have bloated resource methods, so before this is merged, let's see if we have a better option for the language parameter -- especially since it will be set to null the vast majority of the time.

Also, while this is in-progress, flip it back to a draft PR so we don't accidentally merge it before it's ready. Additionally, be sure to squash the commits down before this is marked ready for review -- we don't want the intermediate commits entering the mainline history.

File: src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java
Comment (Januson) [reply src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java:297]: Checkstyle does not support this. If line starts at column 8, it requires all of its continuation lines to start at column 12. Until we move to spotless we are stuck with this format.

Regarding doReturn. While I agree that doReturn is great for readability, I would advise against preferring it over when-then. Mainly it's not type-safe. when-then is less convenient, but does compile-time type checking which saves a ton of work when refactoring.

Comment (Ceiu) [reply src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java:297]: I'd much rather have a `@SuppressWarnings("indentation")` than awful looking syntax, but as far as the doReturn/when thing goes, neither method is "safe" enough to justify only using one or the other. Both have trade-offs [1] in terms of how they function or how they read, and being able to use both is valuable for cases like this, where the actual setup we use isn't significant, and we have something else driving a largely unnecessary code change.

As a slight aside, while a compile-time failure *is* nice, if such an unchecked error hits a PR, it suggests that the test wasn't actually run before the PR was created.

[1] `when` behaves strangely with certain mocks/spys as it's making an actual call to the underlying mocked method during this setup call. Compatibility with the answer syntax and sequenced returns is also different between the two.

File: spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java
Comment (Ceiu) [reply spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java:60]: The admin client might be stateless, but the other clients most certainly are not, as they're tied to a specific user, owner, consumer, or cert. And in the admin client case, there should be almost no resource utilization behind spinning it up, as it should be little more than a set of wrappers around serialization and TCP comms.

However, the my larger concern is:
> a test should be fully isolated.

The more we throw stuff into a `beforeEach` or `beforeAll`, the more tests get chunky, flaky, and annoying to maintain, especially in some cases where we want or expect the state to be pristine, but have inherited stuff from the global setup. In fact, someone is going to run into exactly this problem while we port some of the spec tests, as there is at least one suite of tests which have the "wrong" count of objects for the tests because of inherited state from a `before(:each)` block.

It's a paradigm we should avoid, and discourage over callable setup/utility methods where possible and appropriate.

File: api/candlepin-api-spec.yaml
Comment (bakajstep) [reply api/candlepin-api-spec.yaml]: done

File: spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java
Comment (bakajstep) [reply spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java]: done

File: spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java
Comment (bakajstep) [reply spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java]: done

File: spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java
Comment (bakajstep) [reply spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java]: done

File: spec-tests/src/test/java/org/candlepin/spec/consumers/LocalizationSpecTest.java
Comment (bakajstep) [reply spec-tests/src/test/java/org/candlepin/spec/consumers/LocalizationSpecTest.java]: done

File: spec-tests/src/test/java/org/candlepin/spec/consumers/LocalizationSpecTest.java
Comment (bakajstep) [reply spec-tests/src/test/java/org/candlepin/spec/consumers/LocalizationSpecTest.java]: done

File: src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java
Comment (bakajstep) [reply src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java]: i know but there was problem with checkstyle so i reformat it different way. done

File: src/main/java/org/candlepin/resource/ConsumerResource.java
Comment (nikosmoum) [reply src/main/java/org/candlepin/resource/ConsumerResource.java:664]: With the new generic Requests API (https://github.com/candlepin/candlepin/pull/3527), you can create generic requests and set the language header there, and revert the spec file change

**Review by nikosmoum — CHANGES_REQUESTED** [review]
I haven't reviewed this in detail, but the my main suggestion is to split the the DTO package name changes to it's own separate PR, and get that reviewed/tested/merged separately. These changes are not related to each other which makes it hard to review and not great for the git log history

Right now you have tried to split this into 2 commits, but each commit is mixing changes that should belong to the other (for example the `Add packages to distinguish classpath` commit has most of the package name changes but also includes the deletion of the ruby spec files, while the 'migrate spec test' commit also includes some package name changes)

File: src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java
Comment (Januson) [reply src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java:297]: Good point. suppression will let us format code better until we can get formatter to do that for us.

Running the PR on Jenkins is not an issue I was trying to point out. With `doReturn`, when we refactor some code. Like type of some param. We won't know whether it broke something until we run the tests. With `when.thenReturn`, we will know immediately because the code won't compile. Yes, there are places where when is not a suitable choice, but that does not mean it shouldn't be used at all.

File: spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java
Comment (Januson) [reply spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java]: I agree with you. `beforeAll/beforeEach` should be kept minimal. Big setup block only make reasoning about the tests harder. There are of course cases where some share setup is justified. Like those clients. They are immutable and thread safe. We gain nothing but inconvenience by inlining them into tests.

**Review by nikosmoum — CHANGES_REQUESTED** [review]
A few small things to change, plus I see some of the feedback from @Ceiu hasn't been addressed yet (e.g. the BeforeAll/BeforeEach thing) before we can merge

File: spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/ApiClients.java
Comment (bakajstep) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/ApiClients.java]: done

File: spec-tests/src/main/java/org/candlepin/spec/bootstrap/assertions/CapabilityEnable.java
Comment (bakajstep) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/assertions/CapabilityEnable.java]: done

File: spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java
Comment (bakajstep) [reply spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java:48]: done

File: spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java
Comment (bakajstep) [reply spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java]: done

**Review by nikosmoum — APPROVED** [review]
Looks good to me; I'll let @Ceiu re-review & merge

File: spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java
Comment (nikosmoum) [inline spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java:50]: Let's remove all the @DisplayName annotations since we decided we don't need them

Comment (bakajstep) [reply spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java]: done

**Pull request merged**
